src/Features/Core/Portable/UseThrowExpression/AbstractUseThrowExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,311 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Reflection;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Semantics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.UseThrowExpression
+{
+    /// <summary>
+    /// Looks for patterns of the form:
+    /// <code>
+    /// if (a == null) {
+    ///   throw SomeException();
+    /// }
+    /// 
+    /// x = a;
+    /// </code>
+    /// 
+    /// and offers to change it to
+    /// 
+    /// <code>
+    /// x = a ?? throw SomeException();
+    /// </code>
+    /// 
+    /// Note: this analyzer can be udpated to run on VB once VB supports 'throw' 
+    /// expressions as well.
+    /// </summary>
+    internal abstract class AbstractUseThrowExpressionDiagnosticAnalyzer : DiagnosticAnalyzer, IBuiltInAnalyzer
+    {
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(FeaturesResources.Use_throw_expression), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(FeaturesResources.Use_throw_expression), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+
+        private static DiagnosticDescriptor s_descriptor = 
+            CreateDescriptor(DiagnosticSeverity.Hidden);
+
+        private static DiagnosticDescriptor s_unnecessaryCodeDescriptor =
+            CreateDescriptor(DiagnosticSeverity.Hidden, customTags: DiagnosticCustomTags.Unnecessary);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(s_descriptor, s_unnecessaryCodeDescriptor);
+
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        public bool OpenFileOnly(Workspace workspace) => false;
+
+        private static MethodInfo s_registerOperationActionInfo =
+            typeof(AnalysisContext).GetTypeInfo().GetDeclaredMethod("RegisterOperationActionImmutableArrayInternal");
```

Comment:
Could you add a comment explaining why we need to use reflection instead of public APIs?